Add linear interpolation of tabulated fluid properties, and v,e variable set

spaubt2 opened the pull request:
Refs #20101 

Created BidimensionalInterpolation source and header file. This holds syntax that is the same between Bicubic and Bilinear Interpolation. Added sampleDerivative to Bilinear Interpolation to calculate derivatives at points. 

TFP inherits from TabulatedBicubicFluidProperties and TabulatedBilinearFluidProperties. User can choose desired type of interpolation in the input file. 

Added regression tests to TFP for BilinearInterpolation.

Updated TFP.md file to explain difference in Bicubic/Bilinear and new v,e capabilities 

The unit testing is held up because of templating issues. We can try again once gtest contrib is updated

Commit:
Allow TabulatedFluidProperties to evaluate x_from_v_e functions

Commit:
Testing fluid property lookup in terms of v and e

Commit:
Enable computing e_from_v_h

Commit:
Implement AD interfaces (#20101)

spaubt2 committed:
mplement Bilinear Interpolation Table For Fluid Properties refs #20101

spaubt2 committed:
Use TabulatedFluidProperties base class for bilinear and bicubic

spaubt2 committed:
Fix compilation warnings wrt initialization of temporatry variables, fix docstrings for fluid property routines in TFP, adjust header selection in source file for TFP
refs #20101

spaubt2 committed:
rewrote header and source files of Bi*Interpolation to have syntax match eachother for Bidimensional, Bilinear, and Bicubic. Allowed for adaptation of inheritance system where Bidimensional is the parent class of the other two.
refs #20101

spaubt2 committed:
New names for routines specific to interpolation type oin TabulatedFP
Refs #20101

spaubt2 committed:
Update tabulatedFP documentation to new ve & linear capability, refs #20101

spaubt2 committed:
Add regression testing to new Tabuleated ve linear capability, refs #20101

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1078135/) on 744c9f8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d8857424f6409baa438ce06b4d4a9fa31f21a7b8`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1078616/) on 4733df2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d8857424f6409baa438ce06b4d4a9fa31f21a7b8`

spaubt2 committed:
Add routines to TabulatedFP for T_from_prho s_from_ve pt_from_hs
refs #20101

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1084366/) on e68ce4b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 41e7a8fdd00508719d8d5aa7621bfef4c37e53f7`

spaubt2 committed:
Added routines to SinglePhaseFluidProperties using NewtonMethod to go from v,e to p,T. These routines get used in TFP as well. Modified TabulatedBicubicFluidProperties and TabulatedBilinearFluidProperties to change how interpolate data to change variable sets. Now uses 2D Newton Method to convert between variable sets. TFP WIP still ref #20101

spaubt2 committed:
Created util that does NewtonMethod in 1D and 2D in the fluid_properties module. ref #20101

spaubt2 committed:
fixed e_from_p_rho to get the correct derivatives. Implemented derivatives for T(p,rho) using finite difference method. Added feature to TBicubicFP to allow user to choose whether to get p and T from v,e or v,h or choose to use both.

spaubt2 committed:
Modifed TabulatedBilinearFluidProperties to include ability to choose between from_ve, from_vh, or use both to build tabulated data ref #20101

spaubt2 committed:
Added checkNaNs routine to TabBicubicFP and TabBilinearFP. If nans exist, set variable to be constant as min or max, depending on where nans exist. Also added a checkOutofBounds routine to set variable to be constant at min or max depending on where variable is out of bounds. Added T_from_p_h routine using Newton Method. Added s_from_p_h routine which uses T_from_p_h, then finds s_from_p_T. Added feature to TabBicubicFP and TabBilinearFP which allows user to choose to generate (p,T) from both (v,e) and (v,h) or choose one or the other. Added routine to TFP to check the initial guesses for p and T that are used in Newton's Mehotd. If these guesses are outside the range for max/min of p and T, routine produces an error and informs the user. Ref #20101

spaubt2 committed:
changed fp to a pointer. This allowed for the ability to choose not to generate fluid properties from _fp. ref #20101

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1100223/) on 8d328d2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a4d6ceb558ee7c425777f208ac92fc54242cb9ec`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1100223/) on 8d328d2 wanted to post the following:

The following commit(s) contain fixups:

- 6b5f3d6f2b: fixup! Added checkNaNs routine to TabBicubicFP and TabBilinearFP. If nans exist, set variable to be constant as min or max, depending on where nans exist. Also added a checkOutofBounds routine to set variable to be constant at min or max depending on where variable is out of bounds. Added T_from_p_h routine using Newton Method. Added s_from_p_h routine which uses T_from_p_h, then finds s_from_p_T. Added feature to TabBicubicFP and TabBilinearFP which allows user to choose to generate (p,T) from both (v,e) and (v,h) or choose one or the other. Added routine to TFP to check the initial guesses for p and T that are used in Newton's Mehotd. If these guesses are outside the range for max/min of p and T, routine produces an error and informs the user. Ref #20101

spaubt2 committed:
Format function declarations

spaubt2 committed:
Fixed bug in TFP where if statement for constructing pT_from_ve missing ../../moose/modules/navier_stokes/navier_stokes-devel -i functorfluidpropsVE.i Edited documentation in TFP.md file ref #20101

spaubt2 committed:
Added FluidPropertiesMaterialPH and FluidPropertiesMaterialVE. When used with input files, these allow user to get properties from (v,e) and (p,h). ref #20101

spaubt2 committed:
Resolved issue where fluid properties are initialized twice, reading the csv file twice, causing them to error.

TFP can now run without dependence on another fp object, a csv file with data is all that is needed for this method to work

When variable set inversion of (p,T) from (v,e) or (v,h) occurs while using csv file, it is possible there will be (p,T) values that are out of bounds (converting from a (v,e) or (v,h) grid). When this is the case, pressure and temperature are reset to their respective range min or max values depending on the situation.

ref #20101

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1108101/) on 66e23e2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21301/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a4202606d9021b516117f39241469b352761790f`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1108101/) on 66e23e2 wanted to post the following:

The following commit(s) contain fixups:

- 942713309e: fixup! fixed e_from_p_rho to get the correct derivatives. Implemented derivatives for T(p,rho) using finite difference method. Added feature to TBicubicFP to allow user to choose whether to get p and T from v,e or v,h or choose to use both.

spaubt2 committed:
modified outputWarnings routine in TabulatedBi*FluidProperties.

Now outputs all warnings during variable set inversion at once to provide cleaner look.

ref # 20101

spaubt2 committed:
Resolved issue where fluid properties are initialized twice, reading the csv file twice, causing them to error.

spaubt2 committed:
Modified checkNaNs routine.

Previously was setting temperature and pressure values in the same way based on the specific volume. Should be set in opposite ways due to relation to volume.

ref #20101

spaubt2 committed:
Updated TabulatedFluidProperties.md file to document changes made to variable set inversions that now use Newton's Method, as well as implementaion of a few new routines from older commits. (e_from_p_rho, T_from_p_rho, etc.) ref #20101

spaubt2 committed:
Worked on passing regression tests with TFP.

First time doing this with new capabilities since ___.

WIP, bilinear interpolation is having some issues with getting NaNs in sampleValueandDerivatives routine.

ref #20101

spaubt2 committed:
Worked on implementing SampleValueAndDerivatives to BilinearInterpolation. Lots of special cases were needed. WIP ref #20101

spaubt2 committed:
Worked on fixing derivatives for BilinearInterpolation WIP ref#20101

spaubt2 committed:
created unit tests for BilinearInterpolation utility and NewtonInversion utility WIP ref #20101

spaubt2 committed:
Worked on unit test for BilinearInterpolation and NewtonInversion ref #20101

spaubt2 committed:
Modified BilnearInterpolation.C to obtain derivatives when x1=x2 and y1=y2 along the tob and bottom boundaries

ref #20101

spaubt2 committed:
Removed comments to console used while developing sampleDerivative routine

ref #20101

veeshy commented:
Any timeline for the `gtest` contrib update so this PR could get into the main branch? Quite nice PR!

GiudGiud commented:
I ll work on it this coming week. We ll probably do a separate PR for gtest

GiudGiud commented:
Superseded by https://github.com/idaholab/moose/pull/22060
which is merged now. 

@veeshy fyi
please reach out if you encounter any issue using it.
